pkg/spire/spire.go

```diff
@@ -254,3 +284,206 @@ func (sc *SpireServerApiClient) Close() {
+	trust, err := getTrustBundle(sc.workloadAPI, sc.ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := verifyManifest(resultMap); err != nil {
+		return err
+	}
+
+	if err := verifyCertURI(cert, tr, sc.config.TrustDomain); err != nil {
+		return err
+	}
+
+	if err := verifyCertificateTrust(cert, trust); err != nil {
+		return err
+	}
+
+	for key, _ := range resultMap {
+		if strings.HasSuffix(key, ".sig") {
+			continue
+		}
+		if key == "SVID" {
+			continue
+		}
+		if err := verifyOne(cert.PublicKey, key, resultMap); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (sc *SpireServerApiClient) AppendStatusAnnotation(tr *v1beta1.TaskRun) error {
+	// Add status hash
+	current, err := hashTaskrunStatus(tr)
+	if err != nil {
+		return err
+	}
+	tr.Annotations[TaskRunStatusHashAnnotation] = current
+
+	// Sign with controller private key
+	xsvid, err := sc.fetchSVID()
+	if err != nil {
+		return err
+	}
+
+	s, err := signWithKey(xsvid, current)
+	if err != nil {
+		return err
+	}
+	tr.Annotations[taskRunStatusHashSigAnnotation] = base64.StdEncoding.EncodeToString(s)
+
+	// Store Controller SVID
+	p := pem.EncodeToMemory(&pem.Block{
+		Bytes: xsvid.Certificates[0].Raw,
+		Type:  "CERTIFICATE",
+	})
+	tr.Annotations[controllerSvidAnnotation] = string(p)
+	return nil
+
+}
+
+func hashTaskrunStatus(tr *v1beta1.TaskRun) (string, error) {
+	s, err := json.Marshal(tr.Status)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%x", sha256.Sum256(s)), nil
+}
+
+func CheckStatusAnnotationHash(tr *v1beta1.TaskRun) error {
+	// get stored hash of status
+	hash := tr.Annotations[TaskRunStatusHashAnnotation]
+	// get current hash of status
+	current, err := hashTaskrunStatus(tr)
+	if err != nil {
+		return err
+	}
+	if hash != current {
+		return fmt.Errorf("current status hash and stored annotation hash does not match! Annotation Hash: %s, Current Status Hash: %s", hash, current)
+	}
+	return nil
+}
+
+func getSVID(resultMap map[string]v1beta1.TaskRunResult) (*x509.Certificate, error) {
+	svid, ok := resultMap["SVID"]
+	if !ok {
+		return nil, errors.New("no SVID found")
+	}
+	block, _ := pem.Decode([]byte(svid.Value))
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("invalid SVID: %s", err)
+	}
+	return cert, nil
+}
+
+func getTrustBundle(client *workloadapi.Client, ctx context.Context) (*x509.CertPool, error) {
+	x509set, err := client.FetchX509Bundles(ctx)
+	if err != nil {
+		return nil, err
+	}
+	x509Bundle := x509set.Bundles()
+	if err != nil {
+		return nil, err
+	}
+	trustPool := x509.NewCertPool()
+	for _, c := range x509Bundle[0].X509Authorities() {
+		trustPool.AddCert(c)
+	}
+	return trustPool, nil
+}
+
+func verifyCertURI(cert *x509.Certificate, tr *v1beta1.TaskRun, trustDomain string) error {
+	// URI:spiffe://example.org/ns/default/taskrun/cache-image-pipelinerun-r4r22-fetch-from-git
+	path := "/ns/" + tr.Namespace + "/taskrun/" + tr.Name
+	if cert.URIs[0].Host != trustDomain {
```

Comment:
probably should check just in case the cert is malformed and doesn't have URIs (don't know if its possible), but in case, don't want to segfault

Reply:
check added.